linkerd/meshtls/boring/src/client.rs

```diff
@@ -0,0 +1,163 @@
+use boring::ssl;
+use linkerd_error::Result;
+use linkerd_identity::Name;
+use linkerd_io as io;
+use linkerd_stack::{NewService, Service};
+use linkerd_tls::{
+    client::AlpnProtocols, ClientTls, HasNegotiatedProtocol, NegotiatedProtocolRef, ServerId,
+};
+use std::{future::Future, pin::Pin};
+use tokio::sync::watch;
+
+#[derive(Clone)]
+pub struct NewClient(watch::Receiver<ssl::SslConnector>);
+
+#[derive(Clone)]
+pub struct Connect {
+    server_id: Name,
+    connector: ssl::SslConnector,
+}
+
+pub type ConnectFuture<I> = Pin<Box<dyn Future<Output = io::Result<ClientIo<I>>> + Send>>;
+
+#[derive(Debug)]
+pub struct ClientIo<I>(tokio_boring::SslStream<I>);
+
+// === impl NewClient ===
+
+impl NewClient {
+    pub(crate) fn new(rx: watch::Receiver<ssl::SslConnector>) -> Self {
+        Self(rx)
+    }
+}
+
+impl NewService<ClientTls> for NewClient {
+    type Service = Connect;
+
+    fn new_service(&self, target: ClientTls) -> Self::Service {
+        Connect::new(target, (*self.0.borrow()).clone())
+    }
+}
+
+impl std::fmt::Debug for NewClient {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NewClient").finish()
+    }
+}
+
+// === impl Connect ===
+
+impl Connect {
+    pub(crate) fn new(client_tls: ClientTls, connector: ssl::SslConnector) -> Self {
+        let ServerId(server_id) = client_tls.server_id;
+
+        if let Some(AlpnProtocols(protocols)) = client_tls.alpn {
+            if !protocols.is_empty() {
+                todo!("support ALPN")
+            }
+        }
+
+        Self {
+            server_id,
+            connector,
+        }
+    }
+}
+
+impl<I> Service<I> for Connect
+where
+    I: io::AsyncRead + io::AsyncWrite + Send + Sync + Unpin + 'static,
+{
+    type Response = ClientIo<I>;
+    type Error = io::Error;
+    type Future = ConnectFuture<I>;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        std::task::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        let conn = self.connector.clone();
+        let id = self.server_id.clone();
+        Box::pin(async move {
+            let config = conn
+                .configure()
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+            let io = tokio_boring::connect(config, id.as_str(), io)
+                .await
+                .map_err(|e| match e.as_io_error() {
+                    Some(ioe) => io::Error::new(ioe.kind(), ioe.to_string()),
+                    None => io::Error::new(io::ErrorKind::Other, "unexpected TLS error"),
```

Comment:
this looks like it completely discards any error information if the error _isn't_ an IO error, which could be useful to have --- maybe this should wrap the `HandshakeError` as the inner error's source, so that we still print its message?
```suggestion
                    None => io::Error::new(io::ErrorKind::Other, e)
```

Reply:
boring's error implementation requires the `I: std::fmt::Debug`, which is a bit heavyweight.. I'll add a note about that.

Reply:
ohh, because it contains the IO stream in the error, presumably so that you can get the stream out if you want to use it without TLS. hmm... it would be _really_ nice if they had some way to separate the actual error from the returned IO stream.  but, this is fine, then, given the API constraints!

Reply:
i went ahead and opened https://github.com/cloudflare/boring/pull/57 upstream, which should hopefully help with this as well.

Reply:
We'll at least be able to use `to_string()` here https://github.com/cloudflare/boring/commit/c4e1f474973744b32579797c1ff51335002a4491